src/multibody/numdiff/cost.cpp

```diff
@@ -0,0 +1,112 @@
+///////////////////////////////////////////////////////////////////////////////
+// BSD 3-Clause License
+//
+// Copyright (C) 2018-2020, LAAS-CNRS, New York University,
+//                          Max Planck Gesellschaft
+// Copyright note valid unless otherwise stated in individual files.
+// All rights reserved.
+///////////////////////////////////////////////////////////////////////////////
+
+#include "crocoddyl/core/utils/exception.hpp"
+#include "crocoddyl/multibody/numdiff/cost.hpp"
+
+namespace crocoddyl {
+
+CostNumDiffModel::CostNumDiffModel(const boost::shared_ptr<CostModelAbstract>& model)
+    : CostModelAbstract(model->get_state(), model->get_activation(), model->get_nu(), model->get_with_residuals()),
+      model_(model) {
+  disturbance_ = std::sqrt(2.0 * std::numeric_limits<double>::epsilon());
+}
+
+CostNumDiffModel::~CostNumDiffModel() {}
+
+void CostNumDiffModel::calc(const boost::shared_ptr<CostDataAbstract>& data,
+                            const Eigen::Ref<const Eigen::VectorXd>& x, const Eigen::Ref<const Eigen::VectorXd>& u) {
+  boost::shared_ptr<CostDataNumDiff> data_nd = boost::static_pointer_cast<CostDataNumDiff>(data);
+  data_nd->data_0->cost = 0.0;
+  model_->calc(data_nd->data_0, x, u);
+  data_nd->cost = data_nd->data_0->cost;
+}
+
+void CostNumDiffModel::calcDiff(const boost::shared_ptr<CostDataAbstract>& data,
+                                const Eigen::Ref<const Eigen::VectorXd>& x, const Eigen::Ref<const Eigen::VectorXd>& u,
+                                const bool& recalc) {
+  boost::shared_ptr<CostDataNumDiff> data_nd = boost::static_pointer_cast<CostDataNumDiff>(data);
+
+  if (recalc) {
+    model_->calc(data_nd->data_0, x, u);
+  }
+
+  const double& c0 = data_nd->data_0->cost;
+  data_nd->cost = c0;
+
+  assertStableStateFD(x);
+
+  // Computing the d cost(x,u) / dx
+  data_nd->dx.setZero();
+  for (std::size_t ix = 0; ix < state_->get_ndx(); ++ix) {
+    // x + dx
+    data_nd->dx(ix) = disturbance_;
+    model_->get_state()->integrate(x, data_nd->dx, data_nd->xp);
+    // call the update function on the pinocchio data
+    for (size_t i = 0; i < reevals_.size(); ++i) {
+      reevals_[i](data_nd->xp);
```

Comment:
Are you calling computeAllPinocchio here? If so, could we create a shared pointer of it? The idea is to make cost-numdiff independent of Pinocchio.